lib/AFP/AbstractFilePath/Internal/Common.hs

```diff
@@ -0,0 +1,692 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+-- This template expects CPP definitions for:
+--     MODULE_NAME = Posix | Windows
+--     IS_WINDOWS  = False | True
+
+module AFP.AbstractFilePath.Internal.MODULE_NAME where
+
+-- doctest
+import AFP.AbstractFilePath.Internal.Types
+    ()
+
+#ifdef WINDOWS
+import qualified AFP.Data.ByteString.Short.Word16 as BS
+import AFP.Data.Word16
+    ( isLetter
+    , isSpace
+    , toLower
+    , toUpper
+    , _C
+    , _N
+    , _U
+    , _asterisk
+    , _backslash
+    , _bar
+    , _colon
+    , _greater
+    , _less
+    , _nul
+    , _period
+    , _question
+    , _quotedbl
+    , _semicolon
+    , _slash
+    , _underscore
+    )
+import GHC.Word
+    ( Word16 )
+#else
+import qualified AFP.Data.ByteString.Short as BS
+import Data.Word8
+    ( Word8
+    , isLetter
+    , isSpace
+    , toLower
+    , toUpper
+    , _C
+    , _N
+    , _U
+    , _asterisk
+    , _backslash
+    , _bar
+    , _colon
+    , _greater
+    , _less
+    , _nul
+    , _period
+    , _question
+    , _quotedbl
+    , _semicolon
+    , _slash
+    , _underscore
+    )
+#endif
+
+import Control.Arrow
+    ( second )
+import Data.ByteString
+    ( ByteString )
+import Data.ByteString.Short
+    ( ShortByteString )
+import Data.List
+    ( mapAccumL )
+import Data.Maybe
+    ( isJust )
+import Prelude hiding
+    ( Word )
+
+import Data.String
+
+#ifdef WINDOWS
+type Word = Word16
+#else
+type Word = Word8
+#endif
+
+
+#ifdef WINDOWS
```

Comment:
These huge CPPs are unfortunate, but when I moved it inside the doctest comment, it broke completely.